api/src/main/java/jakarta/data/repository/KeysetPageable.java

```diff
@@ -0,0 +1,328 @@
+ * </pre>
+ *
+ * <h2>Page Numbers and Totals</h2>
+ *
+ * <p>Page numbers, total numbers of elements across all pages, and total
+ * count of pages are not accurate when keyset pagination is used and should
+ * not be relied upon.</p>
+ */
+public class KeysetPageable extends Pageable {
+    /**
+     * Direction of keyset pagination.
+     */
+    public static enum Mode {
+        /**
+         * Indicates forward keyset pagination, which follows the
+         * direction of the {@link OrderBy} annotations, <code>Pageable</code>
+         * {@link Pageable#of(long, long, Sort...) Sort} parameters,
+         * or <code>OrderBy</code> name pattern of the repository method.
+         */
+        NEXT,
+
+        /**
+         * Indicates a request for a page in the reverse direction of
+         * the {@link OrderBy} annotations, <code>Pageable</code>
+         * {@link Pageable#of(long, long, Sort...) Sort} parameters,
+         * or <code>OrderBy</code> name pattern of the repository method.
+         * The order of results on each page follows the sort criteria
+         * and is not reversed.
+         */
+        PREVIOUS
+    }
+
+    /**
+     * Represents keyset values, which can be a starting point for
+     * requesting a next or previous page.
+     */
+    public interface Cursor {
+        /**
+         * Returns whether or not the keyset values of this instance
+         * are equal to those of the supplied instance.
+         * Cursor implementation classes must also match to be equal.
+         *
+         * @return true or false.
+         */
+        @Override
+        public boolean equals(Object o);
+
+        /**
+         * Returns the keyset value at the specified position.
+         *
+         * @param  index position (0 is first) of the keyset value to obtain.
+         * @return the keyset value at the specified position.
+         * @throws IndexOutOfBoundsException if the index is negative
+         *         or greater than or equal to the {@link #size}.
+         */
+        public Object getKeysetElement(int index);
+
+        /**
+         * Returns a hash code based on the keyset values.
+         *
+         * @return a hash code based on the keyset values.
+         */
+        @Override
+        public int hashCode();
+
+        /**
+         * Returns the number of values in the keyset.
+         *
+         * @return the number of values in the keyset.
+         */
+        public int size();
+
+        /**
+         * String representation of the keyset cursor, including the number of
+         * key values in the cursor but not the values themselves.
+         *
+         * @return String representation of the keyset cursor.
+         */
+        @Override
+        public String toString();
+    }
+
+    /**
+     * Built-in implementation of Cursor.
+     */
+    static class CursorImpl implements Cursor {
```

Comment:
should be `final`

Reply:
Good catch. I corrected this in https://github.com/jakartaee/data/pull/52/commits/24b8247df31b01ec682495b338621885a3774204 . Please review it.